xds/src/main/java/io/grpc/xds/internal/sds/DynamicSslContextProvider.java

```diff
@@ -0,0 +1,146 @@
+          CertificateValidationContext certificateValidationContext)
+      throws CertificateException, IOException, CertStoreException;
+
+  // this gets called only when requested secrets are ready...
+  protected final void updateSslContext() {
+    try {
+      CertificateValidationContext localCertValidationContext =
+          generateCertificateValidationContext();
+      SslContextBuilder sslContextBuilder = getSslContextBuilder(localCertValidationContext);
+      CommonTlsContext commonTlsContext = getCommonTlsContext();
+      if (commonTlsContext != null && commonTlsContext.getAlpnProtocolsCount() > 0) {
+        List<String> alpnList = commonTlsContext.getAlpnProtocolsList();
+        ApplicationProtocolConfig apn =
+            new ApplicationProtocolConfig(
+                ApplicationProtocolConfig.Protocol.ALPN,
+                ApplicationProtocolConfig.SelectorFailureBehavior.NO_ADVERTISE,
+                ApplicationProtocolConfig.SelectedListenerFailureBehavior.ACCEPT,
+                alpnList);
+        sslContextBuilder.applicationProtocolConfig(apn);
+      }
+      SslContext sslContextCopy = sslContextBuilder.build();
+      sslContext = sslContextCopy;
+      makePendingCallbacks(sslContextCopy);
+    } catch (CertificateException | IOException | CertStoreException e) {
+      logger.log(Level.SEVERE, "exception in updateSslContext", e);
+    }
+  }
+
+  protected final void callPerformCallback(
+      Callback callback, Executor executor, final SslContext sslContextCopy) {
+    performCallback(
+        new SslContextGetter() {
+          @Override
+          public SslContext get() {
+            return sslContextCopy;
+          }
+        },
+        callback,
+        executor);
+  }
+
+  @Override
+  public final void addCallback(Callback callback, Executor executor) {
+    checkNotNull(callback, "callback");
+    checkNotNull(executor, "executor");
+    // if there is a computed sslContext just send it
+    SslContext sslContextCopy = sslContext;
+    if (sslContextCopy != null) {
+      callPerformCallback(callback, executor, sslContextCopy);
+    } else {
+      synchronized (pendingCallbacks) {
+        pendingCallbacks.add(new CallbackPair(callback, executor));
+      }
+    }
+  }
+
+  protected final void makePendingCallbacks(SslContext sslContextCopy) {
+    synchronized (pendingCallbacks) {
+      for (CallbackPair pair : pendingCallbacks) {
+        callPerformCallback(pair.callback, pair.executor, sslContextCopy);
+      }
+      pendingCallbacks.clear();
+    }
+  }
+
+  /** Propagates error to all the callback receivers. */
+  public final void onError(Status error) {
+    synchronized (pendingCallbacks) {
+      for (CallbackPair callbackPair : pendingCallbacks) {
+        callbackPair.callback.onException(error.asException());
+      }
+      pendingCallbacks.clear();
+    }
+  }
+
+  protected static final class CallbackPair {
+    private final Callback callback;
```

Comment:
Seems callback always comes with an executor. Is it possible to add `getExecutor()` method to `Callback` instead of defining `CallbackPair`?

Reply:
yes, good idea. Done